MudChart: Fix culture and perfomance issues

radderz opened the pull request:
## Description
Fixes issues for some users from the v8.4.0 release.

- Fixes #11020
- Fixes #11019
- Fixes #11062
> This was fixed using MudChartParent.ChartOptions.ShowToolTips which is an existing property on the ChartOptions that was originally for HeatMaps, but I have extended all the other charts to also respect this field (which defaults to true).
![image](https://github.com/user-attachments/assets/171e233b-9ff9-4e64-bc73-6498d1c9b407)
- Add DataMarkerTooltipYValueFormat to Series model to allow the user to specify the format for the {Y_VALUE} tag, it defaults to null.

## How Has This Been Tested?
Unit tests and visually

## Type of Changes
<!-- What type of changes does your code introduce? Put an `x` in only one box that applies best: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (fix or improvement to the website or code docs)

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

radderz committed:
enable disabling of tooltips, also handle invariant cultures for chart tooltips as there are issues with comma based number systems.

radderz committed:
Add a culture test for comma number system cultures, I chose "ru" from the original github issue.

radderz committed:
fix whitespacing

radderz committed:
remove redundant null checks and make static

radderz committed:
add a way to add a template format for the data marker YValue

ScarletKuro commented:
another regression https://github.com/MudBlazor/MudBlazor/issues/11043

ScarletKuro commented:
We also had a conversation with @henon about https://github.com/MudBlazor/MudBlazor/pull/10865, and we want to discuss the `mudObserveElementSize` JS API that you added.
From my understanding of the code, it seems you’re using it to observe size changes of the element. However, we already have the `IResizeObserver` service, which serves the same purpose.
I'm not sure if the current `IResizeObserver` covers everything you need for your charts, and I know the API is a bit clunky, it doesn’t support async handling, as it relies on the `SizeChanged` event and `OnResized`, and lacks automatic cleanup of subscriptions, requiring explicit unsubscription.
But even so, it would still be better to reuse the existing `IResizeObserver` and add the necessary improvements to make it work with your feature, rather than duplicating the code. I plan to refactor this service (the API surface) anyway, similar to what I did with https://github.com/MudBlazor/MudBlazor/pull/7098, and that could provide a good starting point.

radderz commented:
/// <summary>
        /// Gets the width of the specified element.
        /// </summary>
        /// <param name="reference">The element reference.</param>
        /// <returns>The width of the element.</returns>
        double GetWidth(ElementReference reference);
        
        this could also possibly remove the bbox interop function for the labels on the ChartToolTip if it supports svg element sizing.

ScarletKuro commented:
> but the auto unsubscribe could still be worth implementing when subscribing as it would mean it is hard for users to do the wrong thing.

As I mentioned, we can extend/improve the `IResizeObserver` and make it automatically unsubscribe, as you implemented, and make `Unobserve` obsolete. To be honest, I’ve wanted to have auto-unsubscription across all services for performance reasons using the `MutationObserver` for quite some time. 

My long-term plan is for each service to have its own provider (for example, a `<ResizeObserverProvider />` that likes in `MainLayout` and passes the `DotNetObjectReference` of the service to the JS part). The subscription and unsubscription of the component would happen automatically based on an HTML attribute, like this:

```html
<div resizeObserver="0f8fad5b-d9cb-469f-a165-70867728950e">...</div>
```

The `MutationObserver` would listen for this, add it to its internal list, and invoke the corresponding `JSInvokable` with the ID back to the .NET service (i.e., the `<ResizeObserverProvider />`). It would then call back the component with that ID. 

This would likely improve performance (https://github.com/MudBlazor/MudBlazor/issues/6269) and simplify the .NET API.

> perhaps that could be an option of the ResizeObserverOptions

Yeah, it could be.

radderz committed:
fix issue where pie charts and donuts don't work with 100% data values, make tooltip go to center of pie if 100% single item, add tests to test for 100% usecase

radderz commented:
> another regression #11043

This is resolved now

radderz committed:
change test to include have 0 values to ensure it works when adding multiple values but 1 is 100%

radderz committed:
change logic

radderz committed:
revert large angle and then fix tests again
change to using string builder for path for pie/donut

radderz committed:
fix issue with rendering where it uses a -1 rect width

radderz committed:
default _boxWidth to the min width.

radderz commented:
I am fixing another couple of issues I have found, so hold off on merging this just yet

ScarletKuro commented on src/MudBlazor/Components/Chart/Charts/TimeSeries.razor:
```diff
@@ -99,24 +99,24 @@
-                .Replace("{{SERIES_NAME}}", series.Name)
-                .Replace("{{X_VALUE}}", _hoveredDataPoint.LabelXValue)
-                .Replace("{{Y_VALUE}}", _hoveredDataPoint.LabelYValue);
+            .Replace("{{SERIES_NAME}}", series.Name)
```
please revert line changes 112-114

ScarletKuro commented on src/MudBlazor/Components/Chart/Charts/TimeSeries.razor:
```diff
@@ -99,24 +99,24 @@
-                .Replace("{{SERIES_NAME}}", series.Name)
-                .Replace("{{X_VALUE}}", _hoveredDataPoint.LabelXValue)
-                .Replace("{{Y_VALUE}}", _hoveredDataPoint.LabelYValue) ?? string.Empty;
+            .Replace("{{SERIES_NAME}}", series.Name)
```
please revert line changes 117-119

radderz committed:
stop empty labelled pie/donut values from being added to legend as blank items
add label extra width property on axis chart options for larger number support for y axis labels
fix some whitespace issues
fix element size not working correctly due to order of events received at server side (order is not guaranteed it seems when on interactive server mode, a bit of lag can mean the size events are processed in the wrong order)

radderz committed:
revert the horizontal start/end space in favour of extra label width option to make that optional.
change the labels to be auto calculated sizing instead of having manual extra width/height
fix the vertical/horizontal spacing
fix the timeseries date label positioning
fix the bar charts positioning (it had bars showing up off the end of the chart)

redo the tests with the changes

radderz commented:
I have done some more fixes:

revert the horizontal start/end space in favour of extra label width option to make that optional.
change the labels to be auto calculated sizing instead of having manual extra width/height
fix the vertical/horizontal spacing
fix the timeseries date label positioning
fix the bar charts positioning (it had bars showing up off the end of the chart)

redo the tests with the changes

radderz committed:
add in readonlys
change comparison

radderz commented:
@ScarletKuro I have resolved the outstanding bugs I am aware of and fixed some existing ones, I found a weird bug when using interactive server render mode where the timing of the notifications on size change can come out of order. I added a timestamp to the event and compare that it is newer than the last one.

You might need to take that into account in your service changes.

radderz committed:
format

radderz committed:
fix bug

radderz committed:
fix bar positioning and label positioning for bar chart, take into account group width and stroke width to make sure the bars are in the charts bounds.

radderz committed:
change tests to match the new positioning of bars

ScarletKuro commented:
If you look at `TimeSeries.razor` the indent shifted, for example line 80, 88, 121
another file to look is `Line.razor`, I guess this is related to the current bug in razor editor where indention will grow everytime there is formatting.

radderz commented:
> If you look at `TimeSeries.razor` the indent shifted, for example line 80, 88, 121 another file to look is `Line.razor`, I guess this is related to the current bug in razor editor where indention will grow everytime there is formatting.

yea that's such a pain in the butt, will fix it

radderz committed:
fix indentation

radderz committed:
fix indent

radderz commented:
@ScarletKuro - should be fixed now

ScarletKuro commented:
I guess this issue should also close https://github.com/MudBlazor/MudBlazor/issues/11062

radderz committed:
Merge branch 'dev' into MudChartTooltipFixes

radderz commented:
I found an issue with the XAxis labels, will figure that out shortly

radderz committed:
fix xaxis label positioning

radderz committed:
Merge branch 'MudChartTooltipFixes' of https://github.com/radderz/MudBlazor into MudChartTooltipFixes

radderz committed:
fix stacked bar horizontal ticks not following MaxNumYAxisTicks

Anu6is commented:
Font size for the custom graphics text needs to be increased
![image](https://github.com/user-attachments/assets/7ad1e7eb-cde6-483b-8250-7bf3568094b8)

These options are all over the place (not uniform)
- sliders using primary and info color
- checkboxes using info and no color
- over all alignment of components (place check boxes and sliders near to each other)
![image](https://github.com/user-attachments/assets/699bdb25-8949-488c-ab4c-83f135d9cd98)

The 1986-04-26 is slightly cut off on Edge (looks ok on Chrome)
![image](https://github.com/user-attachments/assets/46642bbf-73fe-42ca-b21d-391332f442f6)

Issues when chart series is hidden
- Error hovering over data point
- Line color to tooltip color mismatch 
![image](https://github.com/user-attachments/assets/27520fe8-ee87-47ad-a936-0bfa1ddb7b79)
![image](https://github.com/user-attachments/assets/3135c952-52f3-4ad3-9af9-39ae7d94f434)

Timeseries error when series is hidden
![image](https://github.com/user-attachments/assets/3789df0f-9db9-4a63-b291-925b1a1bf79f)

radderz committed:
fix dashboard error on removing a chart

radderz committed:
fix custom graphics example

radderz committed:
fix issues with rendering series that are not supposed to be rendered on line and timeseries

simplify examples to use similar colors and postiions

radderz committed:
fix tooltip index issue for when some series are not visible

radderz commented:
@Anu6is - I have fixed all of them except for the cutoff label, the labels will cutoff when they are super wide (format) if the container doesn't have space with overflow hidden being used to make the auto sizing work (Besides, if you have many ticks on the x axis with wide formats, they overlap terribly). That's why I introduced rotation for the axis so that wider labels can be handled better.

ScarletKuro committed:
nits
